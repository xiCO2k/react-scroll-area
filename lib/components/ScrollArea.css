.outer {
    overflow: hidden;
    position: relative;

    width: 100%;
    height: 100%;
}

.overflow {
    position: relative;
    overflow-y: scroll;
    overflow-x: hidden;

    height: 100%;

    margin-right: -30px;
    padding-right: 30px;
}

.inner {
    position: relative;
}

.inner:before,
.inner:after {
    content: " ";
    display: table;
}

 .inner:after {
    clear: both;
}

.track {
    position: absolute;
    top: 0px; right: 2px;
    width: 7px;
    height: 100%;

    padding: 0;

    border-radius: 4px;

    opacity: 1;

    pointer-events: none;

    transition: opacity .4s linear;
    -webkit-transition: opacity .4s linear;
    -moz-transition: opacity .4s linear;
    -ms-transition: opacity .4s linear;
    -o-transition: opacity .4s linear;
}

.handler {
    position: absolute;
    top: 0px; left: -1px;
    width: 7px; height: 50px;

    background: #d7d7d7;
    border-radius: 4px;

    transition: background .2s linear;
    -webkit-transition: background .4s linear;
    -moz-transition: background .2s linear;
    -ms-transition: background .2s linear;
    -o-transition: background .2s linear;
}

.handlerHover,
.handlerIsDragging {
    background: #909090;
}

.trackHidden {
    opacity: 0;
}
