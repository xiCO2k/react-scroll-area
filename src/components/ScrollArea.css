.outer {
    overflow: hidden;
    position: relative;

    width: 100%;
    height: 100%;
}
