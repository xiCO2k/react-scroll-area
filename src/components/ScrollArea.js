import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import DOMHelper from '../helpers/DOMHelper';
import style from './ScrollArea.css';

export default class ScrollArea extends Component {

    static propTypes = {
        width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        trackHidden: PropTypes.bool,
        trackHideTime: PropTypes.number,
        minHandlerHeight: PropTypes.number,
        trackMargin: PropTypes.number,
        onScroll: PropTypes.func,

        //for testing purpose
        testInnerHeight: PropTypes.number
    };

    static defaultProps = {
        trackVisible: false,
        trackHidden: false,
        trackHideTime: 1000,
        minHandlerHeight: 70,
        trackMargin: 4,

        //for testing purpose
        testInnerHeight: 0
    };

    constructor(props) {
        super(props);

        this.state = {
            trackActive: props.trackVisible,
            handlerHeight: 0,
            innerMargin: -1,
            innerHeight: 0,
            outerHeight: 0,
            isDragging: false
        };
    }

    componentDidMount() {
        this.onResize(true);
    }

    componentDidUpdate() {
        if (this.getInnerHeight() !== this.getInnerHeight(true)) {
            this.onResize();
        }
    }

    componentWillUnmount() {
        clearTimeout(this.scrollTrackVisibleTimeout);
    }

    getOuterWidth() {
        return parseInt(this.props.width || DOMHelper.getWidth(this.refs['outer']), 10);
    }

    getOuterHeight(fromState = false) {
        if (fromState && this.state.outerHeight) {
            return this.state.outerHeight;
        }

        if (process.env.NODE_ENV === 'testing') {
            return parseInt(this.props.height || 0, 10);
        }

        return parseInt(this.props.height || DOMHelper.getHeight(this.refs['outer']), 10);
    }

    getInnerHeight(fromState = false) {
        if (fromState) {
            return this.state.innerHeight;
        }

        if (process.env.NODE_ENV === 'testing') {
            return this.props.testInnerHeight;
        }

        return DOMHelper.getHeight(this.refs['inner']);
    }

    getTrackHeight(fromState = false) {
        return this.getOuterHeight(fromState) - this.props.trackMargin;
    }

    getHandlerRatio(fromState = false) {
        let innerHeight = this.getInnerHeight(fromState),
            trackHeight = this.getTrackHeight(fromState);

        if (!innerHeight || !trackHeight) {
            return 0;
        }

        return innerHeight / trackHeight;
    }

    getHandlerHeight(trackHeight) {
        if (!trackHeight) {
            return 0;
        }

        return Math.min(Math.max(
            this.props.minHandlerHeight,
            Math.round((trackHeight) / this.getHandlerRatio())
        ), trackHeight);
    }

    getMinHandlerHeight() {
        return Math.min(this.getTrackHeight(true), this.props.minHandlerHeight);
    }

    getHandlerTop() {
        let scrollTop = this.getScrollTop(),
            innerHeight = this.getInnerHeight(true),
            trackHeight = this.getTrackHeight(true),
            handlerHeight = this.getHandlerHeight(trackHeight);


        if (handlerHeight === this.getMinHandlerHeight()) {
            if (!trackHeight) {
                throw new TypeError('the trackHeight can\'t be zero');
            }

            return (
                (trackHeight - handlerHeight) *
                (scrollTop / (innerHeight - trackHeight))
            );
        }

        if (!scrollTop) {
            return 0;
        }



        return Math.round(scrollTop / this.getHandlerRatio());
    }

    getInnerMargin() {
        let outer = this.refs['outer'],
            inner = this.refs['inner'];

        if (!inner.offsetWidth || !outer.offsetWidth) {
            return -1;
        }

        return (inner.offsetWidth - outer.offsetWidth) - 1;
    }

    getScrollTop(target = this.refs['overflow']) {
        return target && target.scrollTop || 0;
    }

    isTrackNeedEvents() {
        return !this.props.trackHidden &&
            !this.props.trackVisible &&
            this.getInnerHeight(true) > this.getOuterHeight(true);
    }

    onMouseWheel(event) {
        var scrollTop = this.getScrollTop(),
            innerHeight = this.getInnerHeight(),
            outerHeight = this.getOuterHeight();

        if (!DOMHelper.isChildOf(event.target, this.refs['outer'], true)) {
            return;
        }

        if ((event.deltaY > 0 && scrollTop >= innerHeight - outerHeight) ||
            (event.deltaY < 0 && scrollTop === 0)) {
            event.stopPropagation();
            event.preventDefault();
            return;
        }
    }

    onResize(fromMount = false) {
        let outerHeight = this.getOuterHeight(),
            trackHeight = this.getTrackHeight(),
            handlerHeight = this.getHandlerHeight(trackHeight),
            state = {
                innerHeight: this.getInnerHeight(),
                handlerHeight,
                outerHeight
            };

        if (this.state.innerMargin === -1) {
            state.innerMargin = this.getInnerMargin();
        }

        this.setState(state);
    }

    onScroll() {
        if (!this.isTrackNeedEvents()) {
            this.forceUpdate();
            return;
        }

        if (this.props.onScroll) {
            let data = {
                scrollTop: this.getScrollTop(),
                innerHeight: this.getInnerHeight(),
                outerHeight: this.getOuterHeight()
            };
            data.complete = (data.scrollTop + data.outerHeight) / data.innerHeight;

            this.props.onScroll(data, this);
        }

        this.forceUpdate();
    }

    onMouseEnter() {
        if (!this.isTrackNeedEvents()) {
            return;
        }

        clearTimeout(this.scrollTrackVisibleTimeout);
        this.setState({ trackActive: true });
    }

    onMouseLeave() {
        if (!this.isTrackNeedEvents() || this.state.isDragging) {
            return;
        }

        this.hideTrack();
    }

    onMouseDown(event) {
        let position = DOMHelper.position(this.refs['handler']),
            offset = DOMHelper.offset(this.refs['handler']);

        if (process.env.NODE_ENV === 'testing') {
            offset = { top: 0, left: this.getOuterWidth() - 5 };
        }

        if (event.pageX < offset.left || !this.state.handlerHover) {
            return;
        }

        this.setState({
            isDragging: true,
            trackActive: !this.props.trackHidden,
            startY: event.pageY,
            originalY: position.top
        });

        this.onMouseMoveFn = this.onMouseMove.bind(this);
        this.onMouseUpFn = this.onMouseUp.bind(this);
        window.addEventListener('mousemove', this.onMouseMoveFn, false);
        window.addEventListener('mouseup', this.onMouseUpFn, false);
    }

    onMouseUp(event) {
        this.setState({
            isDragging: false,
            handlerHover: false
        });

        DOMHelper.ignoreSelection();

        if (!DOMHelper.isChildOf(event.target, this.refs['outer']) &&
            this.isTrackNeedEvents()) {
            this.hideTrack();
        }

        window.removeEventListener('mouseup', this.onMouseUpFn);
        window.removeEventListener('mousemove', this.onMouseMoveFn);
    }

    onMouseMove(event) {
        if (!this.state.isDragging) {
            return;
        }

        let deltaY = event.pageY - this.state.startY,
            offsetY = this.state.originalY + deltaY;

        DOMHelper.ignoreSelection();

        this.refs['overflow'].scrollTop = Math.max(Math.min(
            Math.floor(offsetY * this.getHandlerRatio(true)),
            this.getInnerHeight(true) - this.getOuterHeight(true)
        ), 0);
    }

    onMouseMoveHover(event) {
        if (this.state.isDragging || !this.state.trackActive) {
            return;
        }

        let offset = DOMHelper.offset(this.refs['handler']),
            handlerHover = false;

        if (process.env.NODE_ENV === 'testing') {
            offset = { top: 0, left: this.getOuterWidth() - 5};
        }

        if (event.pageX > offset.left) {
            handlerHover = true;
        }

        this.setState({ handlerHover });
    }

    hideTrack() {
        if (!this.props.trackHideTime) {
            this.setState({ trackActive: false });
            return;
        }

        this.scrollTrackVisibleTimeout = _.delay(() => {
            this.setState({ trackActive: false });
        }, this.props.trackHideTime);
    }

    triggerScroll() {
        this.onScroll();

        return this;
    }

    goToPos(scrollTop, duration = 0) {
        let overflow = this.refs['overflow'];

        if (duration) {
            DOMHelper.scrollTo(overflow, scrollTop, duration);
            return;
        }

        overflow.scrollTop = scrollTop;

        return this;
    }

    goToTop(duration = 0) {
        return this.goToPos(0, duration);
    }

    goToBottom(duration = 0) {
        return this.goToPos(this.getInnerHeight(true) - this.getOuterHeight(true), duration);
    }

    getTrackClassNames() {
        let classNames = [style.track];

        if (!this.state.trackActive) {
            classNames.push(style.trackHidden);
        }

        return classNames.join(' ');
    }

    getHandleClassNames() {
        let classNames = [style.handler];

        if (this.state.isDragging) {
            classNames.push(style.handlerIsDragging);
        }

        if (this.state.handlerHover) {
            classNames.push(style.handlerHover);
        }

        return classNames.join(' ');
    }

    render() {
        return (
            <div
                ref='outer'
                style={{
                    width: this.props.width,
                    height: this.props.height
                }}
                className={style.outer}
                onMouseEnter={this.onMouseEnter.bind(this)}
                onMouseLeave={this.onMouseLeave.bind(this)}
                onMouseDown={this.onMouseDown.bind(this)}
                onMouseMove={this.onMouseMoveHover.bind(this)}
            >
                <div
                    ref='overflow'
                    className={style.overflow}
                    onScroll={this.onScroll.bind(this)}
                    onWheel={this.onMouseWheel.bind(this)}
                >
                    <div
                        ref='inner'
                        className={style.inner}
                        style={{marginRight: this.state.innerMargin}}
                    >
                        {this.props.children}
                    </div>
                </div>
                <div
                    ref='track'
                    className={this.getTrackClassNames()}
                    style={{
                        top: this.props.trackMargin / 2,
                        height: this.getTrackHeight(true)
                    }}
                >
                    <div
                        ref='handler'
                        className={this.getHandleClassNames()}
                        style={{
                            height: this.state.handlerHeight,
                            top: this.getHandlerTop()
                        }}
                    />
                </div>
            </div>
        );
    }
}
