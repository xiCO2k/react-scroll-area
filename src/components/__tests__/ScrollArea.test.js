import React from 'react';
import ScrollArea from '../ScrollArea';
import { shallow, mount } from 'enzyme';

describe('rendering', () => {
    it('renders correctly', () => {
        expect(shallow(<ScrollArea />)).toMatchSnapshot();
    });

    it('has the width and height sent by the props', () => {
        const wrapper = mount(<ScrollArea width="100px" height="100px" />),
            style = wrapper.ref('outer').getNode().style;

        expect(style.width).toBe('100px');
        expect(style.height).toBe('100px');
    });
});

describe('interaction', () => {
    it('calls the callback props.onScroll', () => {
        const props = { onScroll: jest.fn(), testInnerHeight: 200 },
            wrapper = mount(<ScrollArea {...props} />);

        wrapper.instance().triggerScroll();

        expect(props.onScroll).toHaveBeenCalledTimes(1);
    });

    it('should scroll to bottom when the goToBottom() is called', () => {
        const props = {
                width: '100px',
                height: '100px',
                testInnerHeight: 200
            },
            bottomPos = props.testInnerHeight - parseInt(props.height, 10),
            wrapper = mount(<ScrollArea {...props} />);

        jest.useFakeTimers();

        wrapper.instance().goToBottom();
        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(bottomPos);

        wrapper.ref('overflow').getNode().scrollTop = 0;
        wrapper.instance().goToBottom(400);

        jest.runTimersToTime(400);

        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(bottomPos);

    });

    it('should scroll to top when the goToTop() is called', () => {
        const props = {
                width: '100px',
                height: '100px',
                testInnerHeight: 200,
                minHandlerHeight: 1,
                trackMargin: 0
            },
            bottomPos = props.testInnerHeight - parseInt(props.height, 10),
            wrapper = mount(<ScrollArea {...props} />);

        jest.useFakeTimers();

        wrapper.ref('overflow').getNode().scrollTop = 100;
        wrapper.instance()
            .triggerScroll()
            .goToTop();

        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(0);

        wrapper.ref('overflow').getNode().scrollTop = 100;
        wrapper.instance()
            .triggerScroll()
            .goToTop(400);

        jest.runTimersToTime(400);

        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(0);
    });

    it('should scroll to "pos" when the goToPos() is called', () => {
        const props = {
                width: '100px',
                height: '100px',
                testInnerHeight: 200
            },
            bottomPos = props.testInnerHeight - parseInt(props.height, 10),
            wrapper = mount(<ScrollArea {...props} />);

        jest.useFakeTimers();

        wrapper.instance().goToPos(100);
        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(100);

        wrapper.instance().goToPos(50, 400);

        jest.runTimersToTime(400);

        expect(wrapper.ref('overflow').getNode().scrollTop).toBe(50);
    });

    describe('track', () => {

        it('shows when mouse enter', () => {
            const wrapper = mount(<ScrollArea
                width="100px"
                height="100px"
                testInnerHeight={200}
            />);

            expect(wrapper.state('trackActive')).toBe(false);

            wrapper.instance().onMouseEnter();

            expect(wrapper.state('trackActive')).toBe(true);
        });

        it('subtracts the height if the props.trackMargin has value', () => {
            const margin = 10,
                wrapper = mount(
                    <ScrollArea
                        width='100px'
                        height='100px'
                        testInnerHeight={100}
                        trackMargin={margin}
                    />
                );

            expect(wrapper.ref('track').getNode().style.height).toBe((100 - margin) + "px");
            expect(wrapper.ref('track').getNode().style.top).toBe((margin / 2) + "px");
        });

        it('should not show if the "outer" is >= than the "inner"', () => {
            const wrapper = mount(<ScrollArea />);

            expect(wrapper.state('trackActive')).toBe(false);

            wrapper.instance().onMouseEnter();

            expect(wrapper.state('trackActive')).toBe(false);
        });

        it('checks if is hidden after props.trackHideTime', () => {
            const wrapper = mount(
                <ScrollArea
                    width="100px"
                    height="100px"
                    testInnerHeight={200}
                    trackHideTime={100}
                />
            );

            jest.useFakeTimers();

            expect(wrapper.state('trackActive')).toBe(false);

            wrapper.instance().onMouseEnter();
            expect(wrapper.state('trackActive')).toBe(true);

            wrapper.instance().onMouseLeave();
            jest.runTimersToTime(100);
            expect(wrapper.state('trackActive')).toBe(false);
        });

        it('should not hide if isDragging is true', () => {
            const wrapper = mount(
                <ScrollArea
                    width="100px"
                    height="100px"
                    testInnerHeight={200}
                    trackHideTime={0}
                />
            );

            //Sets isDragging to True
            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            wrapper.instance().onMouseDown({ pageX: 96 });
            wrapper.instance().onMouseLeave();

            expect(wrapper.state('trackActive')).toBe(true);
        })

        it('not show when mouseEnter if the props.trackHidden is true', () => {
            const wrapper = mount(
                <ScrollArea
                    width="100px"
                    height="100px"
                    testInnerHeight={200}
                    trackHidden={true}
                />
            );

            expect(wrapper.state('trackActive')).toBe(false);

            wrapper.instance().onMouseEnter();

            expect(wrapper.state('trackActive')).toBe(false);
        });

        it('checks if is always visible if the props.trackVisible is true', () => {
            const wrapper = mount(
                <ScrollArea
                    trackVisible={true}
                    trackHideTime={0}
                />
            );

            expect(wrapper.state('trackActive')).toBe(true);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseLeave();

            expect(wrapper.state('trackActive')).toBe(true);
        });

        it('checks if the handle works when scrolling if props.trackVisible is true', () => {
            const wrapper = mount(
                <ScrollArea
                    width="100px"
                    height="100px"
                    testInnerHeight={200}
                    trackVisible={true}
                    trackHideTime={0}
                    minHandlerHeight={1}
                    trackMargin={0}
                />
            );

            wrapper.instance().goToBottom();
            wrapper.instance().triggerScroll();
            expect(wrapper.ref('handler').getNode().style.top).toBe('50px');

        });
    });

    describe('handler', () => {
        const props = {
            width: '100px',
            height: '100px',
            testInnerHeight: 200
        };

        it('sets "handlerHover: true" when the area is hovered', () => {
            const wrapper = mount(<ScrollArea {...props} />);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            expect(wrapper.state("handlerHover")).toBe(true);
        });

        it('sets "isDragging: true" when the area is onMouseDown()', () => {
            const wrapper = mount(<ScrollArea {...props} />);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            wrapper.instance().onMouseDown({ pageX: 96 });
            expect(wrapper.state("isDragging")).toBe(true);
        });

        it('not sets "isDragging: true" when the area is onMouseDown() if track is not visible', () => {
            const wrapper = mount(<ScrollArea {...props} />);

            wrapper.instance().onMouseDown({ pageX: 96 });
            expect(wrapper.state("isDragging")).toBe(false);
        });

        it('changes the scrollTop when isDragging', () => {
            const wrapper = mount(<ScrollArea {...props} />);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            wrapper.instance().onMouseDown({ pageX: 96, pageY: 0 });
            wrapper.instance().onMouseMoveFn({ pageY: 10 });
            expect(wrapper.ref('overflow').getNode().scrollTop).toBe(20);

            wrapper.instance().onMouseMoveFn({ pageY: 5 });
            expect(wrapper.ref('overflow').getNode().scrollTop).toBe(10);

            //Test the minimum
            wrapper.instance().onMouseMoveFn({ pageY: -2 });
            expect(wrapper.ref('overflow').getNode().scrollTop).toBe(0);

            //Test the maximum
            wrapper.instance().onMouseMoveFn({ pageY: 10000 });
            expect(wrapper.ref('overflow').getNode().scrollTop).toBe(props.testInnerHeight - parseInt(props.height, 10));
        });

        it('hides track when onMouseUp on window if isDragging is true', () => {
            const wrapper = mount(<ScrollArea {...props} trackHideTime={0} />);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            wrapper.instance().onMouseDown({ pageX: 96 });
            wrapper.instance().onMouseLeave();
            wrapper.instance().onMouseUpFn({ target: window });

            expect(wrapper.state("trackActive")).toBe(false);
        });

        it('should not hide track when onMouseUp on window with "isDragging: true" if trackVisible is true', () => {
            const wrapper = mount(<ScrollArea {...props} trackVisible={true} trackHideTime={0} />);

            wrapper.instance().onMouseEnter();
            wrapper.instance().onMouseMoveHover({ pageX: 96 });
            wrapper.instance().onMouseDown({ pageX: 96 });
            wrapper.instance().onMouseLeave();
            wrapper.instance().onMouseUpFn({ target: window });

            expect(wrapper.state("trackActive")).toBe(true);
        });

        it('changes the top after scrolling', () => {
            const wrapper = mount(<ScrollArea {...props} />);

            wrapper.ref('overflow').getNode().scrollTop = 100;
            wrapper.instance().onScroll();

            expect(wrapper.ref('handler').getNode().style.top).toBe("25px");

            wrapper.ref('overflow').getNode().scrollTop = 0;
            wrapper.instance().onScroll();

            expect(wrapper.ref('handler').getNode().style.top).toBe("0px");
        });

        it('has the right height based on the inner height', () => {
            const wrapper = mount(
                <ScrollArea
                    {...props}
                    testInnerHeight={100}
                    trackMargin={0}
                />
            );

            expect(wrapper.ref('handler').getNode().style.height).toBe(props.height);
        });

        it('can not have more height than the track height', () => {
            const wrapper = mount(
                <ScrollArea
                    {...props}
                    testInnerHeight={1}
                    trackMargin={50}
                />
            );

            expect(wrapper.ref('handler').getNode().style.height).toBe((parseInt(props.height, 10) - 50) + "px");
        });

        it('limits the minimum height to the value props.minHandlerHeight', () => {
            const minHandlerHeight = 90,
                wrapper = mount(
                    <ScrollArea
                        {...props}
                        testInnerHeight={1000}
                        trackMargin={0}
                        minHandlerHeight={minHandlerHeight}
                    />
                );

            expect(wrapper.ref('handler').getNode().style.height).toBe(minHandlerHeight + "px");
        });

        it('not limits the minimum height to the value props.minHandlerHeight if its bigger than the track height', () => {
            const trackMargin = 100,
                height = 500,
                wrapper = mount(
                <ScrollArea
                    {...props}
                    height={height}
                    testInnerHeight={10000}
                    trackMargin={trackMargin}
                    minHandlerHeight={height}
                />
            );

            expect(wrapper.ref('handler').getNode().style.height).toBe((height - trackMargin) + "px");

            wrapper.ref('overflow').getNode().scrollTop = 250;
            wrapper.instance().triggerScroll();

            expect(wrapper.ref('handler').getNode().style.top).toBe("0px");
        });
    });
});
