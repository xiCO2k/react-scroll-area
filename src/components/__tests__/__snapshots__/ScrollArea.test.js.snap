// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="outer"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <div
      className="overflow"
      onScroll={[Function]}
      onWheel={[Function]}
    >
      <div
        className="inner"
        style={
          Object {
            "marginRight": -1,
          }
        }
      />
    </div>
    <div
      className="track trackHidden"
      style={
        Object {
          "height": -4,
          "top": 2,
        }
      }
    >
      <div
        className="handler"
        style={
          Object {
            "height": 0,
            "top": 0,
          }
        }
      />
    </div>
  </div>,
  "nodes": Array [
    <div
      className="outer"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
    >
      <div
        className="overflow"
        onScroll={[Function]}
        onWheel={[Function]}
      >
        <div
          className="inner"
          style={
            Object {
              "marginRight": -1,
            }
          }
        />
      </div>
      <div
        className="track trackHidden"
        style={
          Object {
            "height": -4,
            "top": 2,
          }
        }
      >
        <div
          className="handler"
          style={
            Object {
              "height": 0,
              "top": 0,
            }
          }
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ScrollArea
        minHandlerHeight={70}
        testInnerHeight={0}
        trackHidden={false}
        trackHideTime={1000}
        trackMargin={4}
        trackVisible={false}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ScrollArea {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "minHandlerHeight": 70,
          "testInnerHeight": 0,
          "trackHidden": false,
          "trackHideTime": 1000,
          "trackMargin": 4,
          "trackVisible": false,
        },
        "refs": Object {},
        "state": Object {
          "handlerHeight": 0,
          "innerHeight": 0,
          "innerMargin": -1,
          "isDragging": false,
          "outerHeight": 0,
          "trackActive": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="outer"
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          style={
            Object {
              "height": undefined,
              "width": undefined,
            }
          }
        >
          <div
            className="overflow"
            onScroll={[Function]}
            onWheel={[Function]}
          >
            <div
              className="inner"
              style={
                Object {
                  "marginRight": -1,
                }
              }
            />
          </div>
          <div
            className="track trackHidden"
            style={
              Object {
                "height": -4,
                "top": 2,
              }
            }
          >
            <div
              className="handler"
              style={
                Object {
                  "height": 0,
                  "top": 0,
                }
              }
            />
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="outer"
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          style={
            Object {
              "height": undefined,
              "width": undefined,
            }
          }
        >
          <div
            className="overflow"
            onScroll={[Function]}
            onWheel={[Function]}
          >
            <div
              className="inner"
              style={
                Object {
                  "marginRight": -1,
                }
              }
            />
          </div>
          <div
            className="track trackHidden"
            style={
              Object {
                "height": -4,
                "top": 2,
              }
            }
          >
            <div
              className="handler"
              style={
                Object {
                  "height": 0,
                  "top": 0,
                }
              }
            />
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ScrollArea
    minHandlerHeight={70}
    testInnerHeight={0}
    trackHidden={false}
    trackHideTime={1000}
    trackMargin={4}
    trackVisible={false}
  />,
}
`;
